### PR TITLE
feat(cloud9): support AWS CodeCommit repository clone on launch

### LINKED ISSUE #8204: feat(cloud9): support AWS CodeCommit repository clone on launch
see the cloudformation sample:
https://twitter.com/darkosubotica/status/1265156073264218112

### Use Case

To clone any existing codecommit repos on environment launch.

### Other

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by pahud**
feat(cloud9): support AWS CodeCommit repository clone on launch 

Add a new `repositories` property to allow users to clone AWS CodeCommit repositories on environment launch.

Closes #8204

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-cloud9/README.md [hunk 1]
```diff
@@ -49,3 +49,17 @@ const c9env = new cloud9.Ec2Environment(this, 'Cloud9Env3', {
 new cdk.CfnOutput(this, 'URL', { value: c9env.ideUrl });
 ```
 
+Specify `repositories` to clone AWS Codecommit repositories into the environment
```
Comment (skinny85) [inline packages/@aws-cdk/aws-cloud9/README.md:52]: Colon at the end of the sentence:

```suggestion
Specify `repositories` to clone AWS Codecommit repositories into the environment:
```

File: packages/@aws-cdk/aws-cloud9/README.md [hunk 2]
```diff
@@ -49,3 +49,17 @@ const c9env = new cloud9.Ec2Environment(this, 'Cloud9Env3', {
...
+new cloud9.Ec2Environment(stack, 'C9Env', {
+  vpc,
+  repositories: [
+    {
+      repository: codecommit.Repository.fromRepositoryName(stack, 'Repo', 'foo'),
+      path: '/foo',
+    },
+  ],
+});
+
```
Comment (skinny85) [inline packages/@aws-cdk/aws-cloud9/README.md:64]: Unnecessary empty line.

File: packages/@aws-cdk/aws-cloud9/lib/environment.ts [hunk 3]
```diff
@@ -61,6 +62,29 @@ export interface Ec2EnvironmentProps {
...
+   *
+   * @default - do not clone any repository
+   */
+  readonly repositories?: IRepository[];
+}
+
+/**
+ * The interface of the codecommit repository for Cloud9 environment
+ */
+export interface IRepository {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-cloud9/lib/environment.ts:77]: Since this interface is what JSII calls a 'struct' (it only contains properties, not methods), it should not have the `I` prefix (also, this name is super confusing with CodeCommit's `IRepository`). How about `Cloud9Repository` instead?

File: packages/@aws-cdk/aws-cloud9/lib/environment.ts [hunk 4]
```diff
@@ -61,6 +62,29 @@ export interface Ec2EnvironmentProps {
...
+  /**
+   * AWS CodeCommit repository
+   */
+  readonly repository: codecommit.IRepository,
+  /**
```
Comment (skinny85) [inline packages/@aws-cdk/aws-cloud9/lib/environment.ts:82]: There should be an empty line before this line.

File: packages/@aws-cdk/aws-cloud9/lib/environment.ts [hunk 5]
```diff
@@ -61,6 +62,29 @@ export interface Ec2EnvironmentProps {
    * @default - no description
    */
   readonly description?: string;
+
+  /**
+   * The AWS CodeCommit repository to be cloned
+   *
+   * @default - do not clone any repository
+   */
+  readonly repositories?: IRepository[];
```
Comment (skinny85) [inline packages/@aws-cdk/aws-cloud9/lib/environment.ts:71]: I wonder: would `clonedRepositories` be a better name?

File: packages/@aws-cdk/aws-cloud9/lib/environment.ts [hunk 6]
```diff
@@ -61,6 +62,29 @@ export interface Ec2EnvironmentProps {
...
+}
+
+/**
+ * The interface of the codecommit repository for Cloud9 environment
+ */
+export interface IRepository {
+  /**
+   * AWS CodeCommit repository
+   */
+  readonly repository: codecommit.IRepository,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-cloud9/lib/environment.ts:81]: 2 things:

1. I would probably call this `codeCommitRepository` (given the type).
2. Is there any chance Cloud9 will support other types of repositories in the future (GitHub comes to mind)? Because if the answer is 'yes', the current design will not support that. There are ways to 'future proof' it a little bit so that it will support more types if necessary - just gauging the need for that here :).

File: packages/@aws-cdk/aws-cloud9/lib/environment.ts [hunk 7]
```diff
@@ -61,6 +62,29 @@ export interface Ec2EnvironmentProps {
...
+  /**
+   * The path within the development environment's default file system location to clone the AWS CodeCommit
+   * repository into. For example, `/REPOSITORY_NAME` would clone the repository into the
+   * `/home/USER_NAME/environment/REPOSITORY_NAME` directory in the environment.
+   */
+  readonly path: string;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-cloud9/lib/environment.ts:87]: `path` seems way too generic here. Maybe `clonePath`?

File: packages/@aws-cdk/aws-cloud9/package.json [hunk 8]
```diff
@@ -72,12 +72,14 @@
   "dependencies": {
     "@aws-cdk/core": "0.0.0",
     "@aws-cdk/aws-ec2": "0.0.0",
+    "@aws-cdk/aws-codecommit": "0.0.0",
```
Comment (skinny85) [inline packages/@aws-cdk/aws-cloud9/package.json:75]: `codecommit` should be above `ec2`.

File: packages/@aws-cdk/aws-cloud9/package.json [hunk 9]
```diff
@@ -72,12 +72,14 @@
...
     "constructs": "^3.0.2"
   },
   "homepage": "https://github.com/aws/aws-cdk",
   "peerDependencies": {
     "@aws-cdk/core": "0.0.0",
     "@aws-cdk/aws-ec2": "0.0.0",
+    "@aws-cdk/aws-codecommit": "0.0.0",
```
Comment (skinny85) [inline packages/@aws-cdk/aws-cloud9/package.json:82]: Same here (`codecommit` should be above `ec2`).

File: packages/@aws-cdk/aws-cloud9/package.json [hunk 10]
```diff
@@ -95,4 +97,4 @@
   "awscdkio": {
     "announce": false
   }
-}
+}
```
Comment (skinny85) [inline packages/@aws-cdk/aws-cloud9/package.json:100]: Don't remove this trailing newline.

File: packages/@aws-cdk/aws-cloud9/test/cloud9.environment.test.ts [hunk 11]
```diff
@@ -66,4 +67,45 @@ test('throw error when subnetSelection not specified and the provided VPC has no
       instanceType: ec2.InstanceType.of(ec2.InstanceClass.C5, ec2.InstanceSize.LARGE),
     });
   }).toThrow(/no subnetSelection specified and no public subnet found in the vpc, please specify subnetSelection/);
+});
+
+test('imoprt codecommit repo', () => {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-cloud9/test/cloud9.environment.test.ts:72]: This test name doesn't seem great... How about something like `can use CodeCommit repositories`?

File: packages/@aws-cdk/aws-cloud9/test/cloud9.environment.test.ts [hunk 12]
```diff
@@ -66,4 +67,45 @@ test('throw error when subnetSelection not specified and the provided VPC has no
...
+    vpc,
+    repositories: [
+      {
+        repository: codecommit.Repository.fromRepositoryName(stack, 'Repo', 'foo'),
+        path: '/foo',
+      },
+    ],
+  });
+  // THEN
+  expectCDK(stack).to(haveResource('AWS::Cloud9::EnvironmentEC2', {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-cloud9/test/cloud9.environment.test.ts:84]: Use `haveResourceLike`, and then the test can be:

```ts
  expectCDK(stack).to(haveResourceLike('AWS::Cloud9::EnvironmentEC2', {
    Repositories: [
      {
        PathComponent: '/foo',
        RepositoryUrl: {
          'Fn::Join': ['', [
            'https://git-codecommit.',
            { Ref: 'AWS::Region' },
            '.',
            { Ref: 'AWS::URLSuffix' },
            '/v1/repos/foo',
          ]],
        },
      },
    ],
  });
```

File: packages/@aws-cdk/aws-cloud9/test/integ.cloud9.ts [hunk 13]
```diff
@@ -6,18 +7,36 @@ export class Cloud9Env extends cdk.Stack {
...
+function getOrCreateVpc(scope: cdk.Construct): ec2.IVpc {
+  // use an existing vpc or create a new one
+  const vpc = scope.node.tryGetContext('use_default_vpc') === '1' ?
+    ec2.Vpc.fromLookup(scope, 'Vpc', { isDefault: true }) :
+    scope.node.tryGetContext('use_vpc_id') ?
+      ec2.Vpc.fromLookup(scope, 'Vpc', { vpcId: scope.node.tryGetContext('use_vpc_id') }) :
+      new ec2.Vpc(scope, 'Vpc', { maxAzs: 3, natGateways: 1 });
+
+  return vpc;
+}
```
Comment (skinny85) [inline packages/@aws-cdk/aws-cloud9/test/integ.cloud9.ts:38]: Please revert these changes. If you want the VPC to be smaller, maybe pass `maxAzs: 1` when creating it? (And/or other options, like `natGateways: 0`, etc.).

File: packages/@aws-cdk/aws-cloud9/lib/environment.ts (hunk 6 above)
Comment (pahud) [reply packages/@aws-cdk/aws-cloud9/lib/environment.ts:81]: Thanks for bringing it up. It's very likely to support other types of repos in the future. Let me propose a new design and please review again then.

Comment (pahud) [reply packages/@aws-cdk/aws-cloud9/lib/environment.ts]: Trying to propose this design. By the time Cloud9 supports other providers such as Github, we should be able to render it accordingly. What do you think?

```ts
/**
 * Repository provider
 */
export enum Cloud9RepositoryProvider {
  /**
   * AWS CodeCommit
   */
  AWS_CODECOMMIT
}

/**
 * The interface of the codecommit repository for Cloud9 environment
 */
export interface Cloud9Repository {
  /**
   * Repository provider
   * @default AWS_CODECOMMIT
   */
  readonly provider?: Cloud9RepositoryProvider;

  /**
   * Repository URL
   */
  readonly url: string;

  /**
   * The path within the development environment's default file system location to clone the AWS CodeCommit
   * repository into. For example, `/REPOSITORY_NAME` would clone the repository into the
   * `/home/USER_NAME/environment/REPOSITORY_NAME` directory in the environment.
   */
  readonly clonePath: string;
}
```

File: packages/@aws-cdk/aws-cloud9/lib/environment.ts [hunk 14]
```diff
@@ -61,6 +61,46 @@ export interface Ec2EnvironmentProps {
...
    */
   readonly description?: string;
+
+  /**
+   * The AWS CodeCommit repository to be cloned
+   *
+   * @default - do not clone any repository
+   */
+  readonly clonedRepositories?: Cloud9Repository[];
+}
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloud9/lib/environment.ts:71]: This feels like it should be a Union-like class:

```ts
clonedRepos: cloud9.CloneRepository.fromCodeCommit(repo, ‘/foo/bar’)
```

Comment (pahud) [reply packages/@aws-cdk/aws-cloud9/lib/environment.ts:71]: @eladb Wow! That's definitely what I want to explore.

Is there any existing sample with this pattern in other constructs? I'd love to give it a try.

Comment (pahud) [reply packages/@aws-cdk/aws-cloud9/lib/environment.ts:71]: Hi @eladb 

Is this something you expect?

```ts

export interface CloneRepositoryConfig {
  repositoryUrl: string;
  clonePath: string;
}

export class CloneRepository {
  public static fromCodeCommit(repository: codecommit.IRepository, path: string): CloneRepositoryConfig {
    return {
      repositoryUrl: repository.repositoryCloneUrlHttp,
      clonePath: path
    };
  }
}

const app = new cdk.App();
const stack = new cdk.Stack(app, 'DemoStack');
const repo = codecommit.Repository.fromRepositoryName(stack, 'CodeCommitRepo', 'foo');

new Ec2Environment(stack, 'C9', {
  vpc: new ec2.Vpc(stack, 'Vpc'),
  clonedRepositories: [ CloneRepository.fromCodeCommit(repo, '/foo/bar') ],
});

```

Comment (eladb) [reply packages/@aws-cdk/aws-cloud9/lib/environment.ts:72]: `ClonedRepositoryConfig` is redundant (typescript uses structural typing so this should work):

```ts
export class CloneRepository {
  public static fromCodeCommit(repository: codecommit.IRepository, path: string): CloneRepository {
    return {
      repositoryUrl: repository.repositoryCloneUrlHttp,
      clonePath: path
    };
  }

  public readonly repositoryUrl: string;
  public readonly clonePath: string;
  
  private constructor() { }
}

const app = new cdk.App();
const stack = new cdk.Stack(app, 'DemoStack');
const repo = codecommit.Repository.fromRepositoryName(stack, 'CodeCommitRepo', 'foo');

new Ec2Environment(stack, 'C9', {
  vpc: new ec2.Vpc(stack, 'Vpc'),
  clonedRepositories: [ CloneRepository.fromCodeCommit(repo, '/foo/bar') ],
});
```

Comment (pahud) [reply packages/@aws-cdk/aws-cloud9/lib/environment.ts:72]: @eladb 

Looks like we'll need to initialize the two properties in the constructor.

圖片

Is it safe simply omit the properties and constructor?

```ts
export class CloneRepository {
  public static fromCodeCommit(repository: codecommit.IRepository, path: string): CloneRepository {
    return {
      repositoryUrl: repository.repositoryCloneUrlHttp,
      clonePath: path
    };
  }
}
```

File: packages/@aws-cdk/aws-cloud9/README.md [hunk 15]
```diff
@@ -49,3 +49,22 @@ const c9env = new cloud9.Ec2Environment(this, 'Cloud9Env3', {
...
+Specify `repositories` to clone AWS Codecommit repositories into the environment:
+
+```ts
+const repo = new codecommit.Repository(this, 'Repo', {
+  repositoryName: 'foo',
+});
+
+const c9env = new cloud9.Ec2Environment(this, 'C9Env', {
+  vpc,
+  clonedRepositories: [
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloud9/README.md:63]: Needs to use CloneRepository

File: packages/@aws-cdk/aws-cloud9/lib/environment.ts (hunk 14 above)
Comment (eladb) [reply packages/@aws-cdk/aws-cloud9/lib/environment.ts:72]: Just add to the ctor:

```ts
class CloneRepository {
  ctor(public readonly x, public readonly y) {}
}
```

File: packages/@aws-cdk/aws-cloud9/README.md [hunk 16]
```diff
@@ -49,3 +49,17 @@ const c9env = new cloud9.Ec2Environment(this, 'Cloud9Env3', {
...
+
+Use `clonedRepositories` to clone AWS Codecommit repositories into the environment:
+
+```ts
+  const repo = codecommit.Repository.fromRepositoryName(stack, 'Repo', 'foo');
+
+  new cloud9.Ec2Environment(stack, 'C9Env', {
+    vpc,
+    clonedRepositories: [
+      cloud9.CloneRepository.fromCodeCommit(repo, '/foo'),
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloud9/README.md:62]: Use more meaningful value instead of `/foo` here to make it clearer what this parameter is. For example `/src`.

File: packages/@aws-cdk/aws-cloud9/lib/environment.ts [hunk 17]
```diff
@@ -125,11 +133,32 @@ export class Ec2Environment extends cdk.Resource implements IEc2Environment {
...
\ No newline at end of file
+}
+
+/**
+ * The class for different repository providers
+ */
+export class CloneRepository {
+  /**
+   * import repository to cloud9 environment from AWS CodeCommit
+   */
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloud9/lib/environment.ts:155]: Missing docs for @params

File: packages/@aws-cdk/aws-cloud9/lib/environment.ts [hunk 18]
```diff
@@ -125,11 +133,32 @@ export class Ec2Environment extends cdk.Resource implements IEc2Environment {
...
+  public static fromCodeCommit(repository: codecommit.IRepository, path: string): CloneRepository {
+    return {
+      repositoryUrl: repository.repositoryCloneUrlHttp,
+      clonePath: path,
+    };
+  }
+
+  private constructor(public readonly repositoryUrl: string, public readonly clonePath: string) {}
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloud9/lib/environment.ts:163]: Use the same names as the ones in cfn (I believe it’s `pathComponent`)

**Comment by pahud** [comment]
Hi @skinny85 

I am ready. Please kindly review again. Thanks.

**Comment by skinny85** [comment]
@pahud before I review, I want to understand something first.

Do you not need any permissions on the CodeCommit repo that you're cloning? Did you try and spin up a new Cloud9 environment with this functionality, and was the repository cloned correctly?

**Comment by pahud** [comment]
> @pahud before I review, I want to understand something first.
> 
> Do you not need any permissions on the CodeCommit repo that you're cloning? Did you try and spin up a new Cloud9 environment with this functionality, and was the repository cloned correctly?

Yes. I have run two tests:

The first one is to create a new codecommit repo and cloud9 environment in the same stack, which is exactly what my integ test does

```ts
import * as codecommit from '@aws-cdk/aws-codecommit';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as cdk from '@aws-cdk/core';
import * as cloud9 from '../lib';

export class Cloud9Env extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const vpc = new ec2.Vpc(this, 'VPC', {
      maxAzs: 2,
      natGateways: 1,
    });

    // create a codecommit repository to clone into the cloud9 environment
    const repo = new codecommit.Repository(this, 'Repo', {
      repositoryName: 'foo',
    });

    // create a cloud9 ec2 environment in a new VPC
    const c9env = new cloud9.Ec2Environment(this, 'C9Env', {
      vpc,
      // clone repositories into the environment
      clonedRepositories: [
        cloud9.CloneRepository.fromCodeCommit(repo, '/foo'),
      ],
    });
    new cdk.CfnOutput(this, 'URL', { value: c9env.ideUrl });
    new cdk.CfnOutput(this, 'ARN', { value: c9env.ec2EnvironmentArn });
  }
}

const app = new cdk.App();

new Cloud9Env(app, 'C9Stack');

```

圖片

The c9 environment and codecommit repo all created correctly and when I click the cloud9 URL:

圖片
It just cloned the new codecommit repo into the target path I specified, though the repo is empty I still can see the `.git` folder inside.

Another test is to create a new cloud9 environment and clone an existing codecommit repository which I created manually.

```ts
    // import an existing codecommit repo
    const repo = codecommit.Repository.fromRepositoryName(this, 'Repo', 'wildrydes-site');

    const c9env = new cloud9.Ec2Environment(this, 'C9Env', {
      vpc,
      // clone repositories into the environment
      clonedRepositories: [
        cloud9.CloneRepository.fromCodeCommit(repo, '/foo'),
      ],
    });
```

圖片


To make it clear, I just updated the README with this sample:


### Cloning Repositories

Use `clonedRepositories` to clone one or multiple AWS Codecommit repositories into the environment:

```ts
// create a codecommit repository to clone into the cloud9 environment
const repoNew = new codecommit.Repository(this, 'RepoNew', {
  repositoryName: 'new-repo',
});

// import an existing codecommit repository to clone into the cloud9 environment
const repoExisting = codecommit.Repository.fromRepositoryName(stack, 'RepoExisting', 'existing-repo');

// create a new Cloud9 environment and clone the two repositories
new cloud9.Ec2Environment(stack, 'C9Env', {
  vpc,
  clonedRepositories: [
    cloud9.CloneRepository.fromCodeCommit(repoNew, '/src/new-repo'),
    cloud9.CloneRepository.fromCodeCommit(repoExisting, '/src/existing-repo'),
  ],
});
```

And both repos were cloned correctly as expected:

圖片
